daphne/src/testing.rs

```diff
@@ -470,32 +474,126 @@ impl DapLeader<BearerToken> for MockAggregator {
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let leader_state_store = leader_state_store_mutex_guard.deref_mut();
+
+        // Store Collect ID and CollectReq into LeaderState.
+        let collect_id = Id(rng.gen());
+        let leader_state = leader_state_store
+            .entry(collect_req.task_id.clone())
+            .or_insert_with(LeaderState::new);
+        leader_state.collect_ids.push_back(collect_id.clone());
+        let collect_job_state = CollectJobState::Pending(collect_req.clone());
+        leader_state
+            .collect_jobs
+            .insert(collect_id.clone(), collect_job_state);
+
+        // Construct a new Collect URI for this CollectReq.
+        let collect_uri = task_config
+            .leader_url
+            .join(&format!(
+                "/collect/task/{}/req/{}",
+                collect_req.task_id.to_base64url(),
+                collect_id.to_base64url(),
+            ))
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+
+        Ok(collect_uri)
     }
 
+    // Called to retrieve completed CollectResp at the request of Collector.
     async fn poll_collect_job(
         &self,
-        _task_id: &Id,
-        _collect_id: &Id,
+        task_id: &Id,
+        collect_id: &Id,
     ) -> Result<DapCollectJob, DapError> {
-        unreachable!("not implemented");
+        let mut leader_state_store_mutex_guard = self
+            .leader_state_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let leader_state_store = leader_state_store_mutex_guard.deref_mut();
+
+        let leader_state = leader_state_store
+            .get(task_id)
+            .ok_or_else(|| DapError::fatal("collect job not found for task_id"))?;
+        if let Some(collect_job_state) = leader_state.collect_jobs.get(collect_id) {
+            match collect_job_state {
+                CollectJobState::Pending(_) => Ok(DapCollectJob::Pending),
+                CollectJobState::Processed(resp) => Ok(DapCollectJob::Done(resp.clone())),
+            }
+        } else {
+            Ok(DapCollectJob::Unknown)
+        }
     }
 
+    // Called to retrieve pending CollectReq.
     async fn get_pending_collect_jobs(
         &self,
-        _task_id: &Id,
+        task_id: &Id,
     ) -> Result<Vec<(Id, CollectReq)>, DapError> {
-        unreachable!("not implemented");
+        let mut leader_state_store_mutex_guard = self
+            .leader_state_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let leader_state_store = leader_state_store_mutex_guard.deref_mut();
+
+        let leader_state = leader_state_store
+            .get_mut(task_id)
+            .ok_or_else(|| DapError::fatal("collect job not found for task_id"))?;
+        let mut res: Vec<(Id, CollectReq)> = Vec::default();
+
+        // Pop collect IDs from the queue and insert them and the associated requests to the response.
```

Comment:
We don't want to remove the collect jobs since we don't necessarily consume them. (Take a look at `DapLeader::process()` to see why.) Instead we just want to copy them.

Reply:
I've moved the deletion process in `finish_collect_job` method.